Source/WebCore/css/CSSCounterStyle.h

```diff
@@ -0,0 +1,177 @@
+    bool isFallbackUnresolved() { return !m_fallbackReference; }
+    bool isExtendsUnresolved() { return m_isExtendedUnresolved; };
+    bool isExtendsSystem() const { return m_system == CounterStyleSystem::Extends; }
+    void extendAndResolve(const CounterStyle&);
+private:
+    // CounterStyle() = default;
+    CounterStyle(const StyleRuleCounterStyle&);
+    // CounterStyle(CSSCounterStyleRule&, CounterStyleManager&, Name, CounterStyleSystem, NegativeSymbols, Symbol, Symbol, Ranges&&, Pad, Name, Vector<Symbol>&&, CounterStyleSpeakAs, Name);
+    CounterStyle(Name, CounterStyleSystem, NegativeSymbols, Symbol, Symbol, Ranges&&, Pad, Name, Vector<Symbol>&&, CounterStyleSpeakAs, Name);
+
+    // https://www.w3.org/TR/css-counter-styles-3/#counter-style-range
+    bool isInRange(int) const;
+    // https://www.w3.org/TR/css-counter-styles-3/#counter-style-negative
+    bool usesNegativeSign();
+    bool shouldApplyNegativeSymbols(int);
+    // https://www.w3.org/TR/css-counter-styles-3/#counter-style-fallback
+    Ref<CounterStyle> fallback();
+    // Generates a CounterStyle object as it was defined by a 'decimal' descriptor. It is used as a last-resource in case we can't resolve fallback references.
+    void applyPadSymbols(String&);
+    void applyNegativeSymbols(String&);
+    // Initial text representation for the counter, before applying pad and/or negative symbols. Suffix and Prefix are also not considered as described by https://www.w3.org/TR/css-counter-styles-3/#counter-styles.
+    String initialRepresentation(int);
+
+    CounterStyleSystem system() const { return m_system; }
+    String counterForSystemCyclic(int);
+    String counterForSystemFixed(int);
+    String counterForSystemSymbolic(int);
+    String counterForSystemAlphabetic(int);
+    String counterForSystemNumeric(int);
+    String counterForSystemAdditive(int);
+
+    bool isPredefinedCounterStyle() const { return m_predefinedCounterStyle; }
+    bool isAutoRange() const { return m_ranges.isEmpty(); }
+    // void extend(const CounterStyle&);
+
+    // Fixed and Extends system can/must have extra data in the system descriptor (https://www.w3.org/TR/css-counter-styles-3/#fixed-system and https://www.w3.org/TR/css-counter-styles-3/#extends-system).
+    void extractDataFromSystemDescriptor(const StyleRuleCounterStyle&);
+
+    bool m_predefinedCounterStyle { false };
+
+    Name m_name;
+    CounterStyleSystem m_system;
+    NegativeSymbols m_negativeSymbols;
+    Symbol m_prefix;
+    Symbol m_suffix;
+    Ranges m_ranges;
+    Pad m_pad;
+    Name m_fallbackName;
+    WeakPtr<const CounterStyle> m_fallbackReference { nullptr };
+    Vector<Symbol> m_symbols;
+    AdditiveSymbols m_additiveSymbols;
+    CounterStyleSpeakAs m_speakAs;
+    Name m_speakAsNameReference;
+    Name m_extendsName;
+    WeakPtr<const CounterStyle> m_extendsReference { nullptr };
+    int m_fixedSystemFirstSymbolValue { 1 };
+
+    OptionSet<ExplicitlySetDescriptors> m_explicitlySetDescriptors;
```

Comment:
Instead of maintaining this OptionSet maybe you could make the individual fields std::optional<>?

Reply:
It doesn't achieve the same intent. All the descriptors values are defined, implicitly or explicitly. No explicit values should not be overridden when resolving Extends. I.E.: when CounterStyleA extends CounterStyleB, A will just inherit the values from the descriptors that were explicitly defined by the author, but all the descriptors have at least an implicit value.